File: Java Solutions/113_Path Sum II.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private void dfs(TreeNode node, List<Integer> path, int remainingSum, List<List<Integer>> res) {
        if (node == null) {
            return;
        }

        path.add(node.val);

        if (node.left == null && node.right == null && remainingSum == node.val) {
            res.add(new ArrayList<>(path));
        } else {
            this.dfs(node.left, path, remainingSum - node.val, res);
            this.dfs(node.right, path, remainingSum - node.val, res);
        }

        path.remove(path.size() - 1);
    }

    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        List<Integer> path = new ArrayList<Integer>();
        this.dfs(root, path, targetSum, res);
        return res;
    }
}

File: Tree/113_Path Sum II.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        res = []

        def dfs(node, path, remaining_sum):
            nonlocal res
            if not node:
                return
            path.append(node.val)
            if not node.left and not node.right and remaining_sum == node.val:
                res.append(path.copy())
            else:
                dfs(node.left, path, remaining_sum - node.val)
                dfs(node.right, path, remaining_sum - node.val)
            path.pop()

        dfs(root, [], targetSum)
        return res